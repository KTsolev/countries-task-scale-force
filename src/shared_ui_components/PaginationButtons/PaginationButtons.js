import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { selectCountries } from '../../redux/selectors';
import { fetchCountries } from '../../redux/countriesSlice';

export const PaginationButtons = (props) => {
  const { itemsPerPage } = props;
  const dispatch = useDispatch();
  const fetchedCountries = useSelector(selectCountries);
  const { pages, page } = fetchedCountries;
  const [buttons, setButtons] = useState([]);

  useEffect(() => {
    calculatePaginationButtons();    
  }, [pages, itemsPerPage]);

  const calculatePaginationButtons = () => {
    let items = [];
    
    for(let p=1; p <= pages; p++) {
      items.push(p);
    }

    setButtons([...items]);
  }

  const onPaginationButtonClick = (event) => {
    const value = event.currentTarget.innerHTML;
    dispatch(fetchCountries({page: value, itemsPerPage, reload: true }));
  }

  return (
    <div className='pagination-controls'>
        {buttons.map(button => <span 
              key={button}
              onClick={onPaginationButtonClick}
              className={button === page ? 'pagination-controls-item pagination-controls-item--current' : 'pagination-controls-item'}
            >
              {button}
            </span>)}
    </div>
  )
}