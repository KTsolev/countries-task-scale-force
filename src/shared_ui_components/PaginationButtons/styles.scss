
    .pagination-controls {
        display: flex;
        justify-content: center;
        align-items: center;
        flex-direction: row;
        position: sticky;
        bottom: 0;
        padding: 6px;
        flex-wrap: wrap;

        &-item {
            display: block;
            margin: 0 2px;
            width: 20px;
            height: 20px; 
            font-size: 14px;
            text-align: center;
            font-weight: 400;
            background-color: #38a1d1;
            color: #ffe4c4;
            border-radius: 40px;
      
            &--current {
                background-color: #6e7f96;
                border: 1px solid #ffe4c4; 
                color: #ffe4c4;
            }

            &:hover {
                @include hover-mixin(#526175, #ffe4c4);
            }
        }
    }