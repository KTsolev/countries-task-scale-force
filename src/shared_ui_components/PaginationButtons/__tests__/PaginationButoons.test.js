/* eslint-disable jest/no-mocks-import */
import React from 'react';
import { PaginationButtons } from '../PaginationButtons';
import renderer from 'react-test-renderer';
import jsonCountries from '../../../__mocks__/mockedParsedJSON';

const mockedCountries = jsonCountries;

jest.mock('react-redux', () => ({
  useSelector: selector => selector(),
  useDispatch: () => jest.fn()
}));

jest.mock('../../../redux/selectors.js', () => ({
    selectCountries: () => mockedCountries
}));

describe('PaginationButtons tests', () => {
    afterEach(() => {
        jest.resetAllMocks();
    });
    
    it('should render component without props', () => {
       const tree = renderer.create(<PaginationButtons />);
        
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should render 6 butttons with curren one number 1 and should handle onclick', () => {
        const tree = renderer.create(<PaginationButtons itemsPerPage={50} />);

        expect(tree.root.findAllByType('span').length).toEqual(6);

        
        expect(tree.root.findAllByProps({className: 'pagination-controls-item pagination-controls-item--current'})[0].props.children).toEqual(1);
        expect(tree.toJSON()).toMatchSnapshot();
        const event = {
            currentTarget: {
                innerHTML: 3
            }
        }

        tree.root.findAllByType('span')[3].props.onClick(event);
        expect(tree.root.findAllByProps({className: 'pagination-controls-item pagination-controls-item--current'})[0].props.children).toEqual(1);
        expect(tree.toJSON()).toMatchSnapshot();
    });
});