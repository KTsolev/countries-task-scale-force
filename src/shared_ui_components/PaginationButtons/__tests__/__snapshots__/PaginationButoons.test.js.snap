// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationButtons tests should render 6 butttons with curren one number 1 and should handle onclick 1`] = `
<div
  className="pagination-controls"
>
  <span
    className="pagination-controls-item pagination-controls-item--current"
    onClick={[Function]}
  >
    1
  </span>
  <span
    className="pagination-controls-item"
    onClick={[Function]}
  >
    2
  </span>
  <span
    className="pagination-controls-item"
    onClick={[Function]}
  >
    3
  </span>
  <span
    className="pagination-controls-item"
    onClick={[Function]}
  >
    4
  </span>
  <span
    className="pagination-controls-item"
    onClick={[Function]}
  >
    5
  </span>
  <span
    className="pagination-controls-item"
    onClick={[Function]}
  >
    6
  </span>
</div>
`;

exports[`PaginationButtons tests should render 6 butttons with curren one number 1 and should handle onclick 2`] = `
<div
  className="pagination-controls"
>
  <span
    className="pagination-controls-item pagination-controls-item--current"
    onClick={[Function]}
  >
    1
  </span>
  <span
    className="pagination-controls-item"
    onClick={[Function]}
  >
    2
  </span>
  <span
    className="pagination-controls-item"
    onClick={[Function]}
  >
    3
  </span>
  <span
    className="pagination-controls-item"
    onClick={[Function]}
  >
    4
  </span>
  <span
    className="pagination-controls-item"
    onClick={[Function]}
  >
    5
  </span>
  <span
    className="pagination-controls-item"
    onClick={[Function]}
  >
    6
  </span>
</div>
`;

exports[`PaginationButtons tests should render component without props 1`] = `
<div
  className="pagination-controls"
/>
`;
